Internal: Propagate headers & contexts to sub-requests

spinscale opened the pull request:
Whenever a query parser (or any other component) issues another
request as part of a request, the headers and the context has to
be supplied as well.

In order to do this, the `SearchContext` has to have those headers
available, which in turn means, the shard level request needs to
copy those from the original `SearchRequest`

This commit introduces two new interface to supply the needed methods
to work with context and headers.

Closes #10979

spinscale commented:
@uboness can you take a look, so we can validate this approach

The PR is not yet finished completely, needs some cleaning up and moving packages around.

uboness commented:
@spinscale sure... Will do it today

uboness commented on src/main/java/org/elasticsearch/common/ContextSupport.java:
```diff
@@ -0,0 +1,36 @@
+import com.carrotsearch.hppc.ObjectObjectAssociativeContainer;
+import org.elasticsearch.common.collect.ImmutableOpenMap;
+
+public interface ContextSupport {
```
can we rename this to `HasContext`?

uboness commented on src/main/java/org/elasticsearch/common/HeaderSupport.java:
```diff
@@ -0,0 +1,47 @@
+/**
+ *
+ */
+public interface HeaderSupport {
```
can we rename this to `HasHeaders`

uboness commented on src/main/java/org/elasticsearch/common/HeaderSupport.java:
```diff
@@ -0,0 +1,47 @@
+
+    Set<String> getHeaders();
+
+
```
why not also add `copyHeadersFrom` method (like we do with the context)

uboness commented on src/main/java/org/elasticsearch/search/suggest/SuggestParseElement.java:
```diff
@@ -44,12 +47,15 @@ public SuggestParseElement(Suggesters suggesters) {
     public void parse(XContentParser parser, SearchContext context) throws Exception {
-        SuggestionSearchContext suggestionSearchContext = parseInternal(parser, context.mapperService(), context.shardTarget().index(), context.shardTarget().shardId());
+        SuggestionSearchContext suggestionSearchContext = parseInternal(parser, context.mapperService(), context.shardTarget().index(), context.shardTarget().shardId(), context, context);
         context.suggest(suggestionSearchContext);
```
wondering.. can we move the copy headers/ctx tot he `SearchContex#suggest` method?

spinscale replied:
no need to anymore, we can remove the whole suggest stuff due to https://github.com/elastic/elasticsearch/commit/7efc43db25495b92833c4bc804f604b82597c157

uboness commented on src/main/java/org/elasticsearch/search/suggest/SuggestionSearchContext.java:
```diff
@@ -20,25 +20,34 @@
 /**
  */
-public class SuggestionSearchContext {
+public class SuggestionSearchContext extends ContextHolder implements ContextSupport, HeaderSupport {
```
`ContextHolder` already implements `ContextSupport`

uboness commented on src/main/java/org/elasticsearch/search/suggest/SuggestionSearchContext.java:
```diff
@@ -20,25 +20,34 @@
-    public static class SuggestionContext {
-        
+
+    public static class SuggestionContext extends ContextHolder implements ContextSupport, HeaderSupport {
```
same as above

uboness commented on src/test/java/org/elasticsearch/transport/local/ContextLocalTransportTests.java:
```diff
@@ -0,0 +1,450 @@
+        }
+    }
+
+    public static class LoggingFilter extends ActionFilter.Simple {
```
cool!

uboness commented on src/test/java/org/elasticsearch/transport/local/ContextLocalTransportTests.java:
```diff
@@ -0,0 +1,450 @@
+        @Override
+        protected boolean apply(String action, ActionRequest request, ActionListener listener) {
+//            logger.info("### action [{}], request [{}]", action, request.getClass().getName());
+            if (request instanceof GetRequest) {
```
wondering.. what if this filter would catch all requests (regardless of the type)... and in each test we assert that all requests contain the headers & ctx? At the end of the day, we do want all requests (regardless of type) to contain the headers. Maybe it'll help expose requests that we're missing now.

spinscale replied:
need to try this out for headers, by setting some header in the client() and transportClient() settings (not sure if we propagate those). This will not work at all with the context, because that only lives on the local node and is not copied.. unless we fill the context for every request manually, but then it doesnt make a lot of sense to check if it is configured.

spinscale replied:
tested and this doesnt work. For example an `IndexRequest` into a non-existing index triggers a `PutMappingRequest`, which is not supposed to have those information, we need to be more selective about this...

Same for index and refresh...

However I added more calls to be tested, that those contain the headers in the test, like manual index and refresh requests

spinscale commented:
@uboness handled your review comments, removed the suggester features due to the latest changes by areek, also changed the tests to check for more requests and if they contain headers (indexing, refresh)

jaymode commented on src/main/java/org/elasticsearch/rest/BaseRestHandler.java:
```diff
@@ -66,12 +66,7 @@ public final void handleRequest(RestRequest request, RestChannel channel) throws
-                    actionRequest.putHeader(usefulHeader, headerValue);
-                }
-            }
+            actionRequest.copyHeadersFrom(actionRequest);
```
I think this should be ```actionRequest.copyHeadersFrom(restRequest)``` ?

spinscale replied:
I had to restore the old behaviour, as only specified/registered HTTP headers are included in the real request... added a test as well

uboness commented on src/main/java/org/elasticsearch/common/ContextAndHeaderHolder.java:
```diff
@@ -23,20 +23,21 @@
  *
  */
-public class ContextHolder {
+public class ContextAndHeaderHolder<T> implements HasContext, HasHeaders {
```
shouldn't this just implement `HasContextAndHeaders`?

spinscale replied:
fixed

uboness commented on src/main/java/org/elasticsearch/common/HasContextAndHeaders.java:
```diff
@@ -0,0 +1,26 @@
+/**
+ * marker interface
+ */
+public interface HasContextAndHeaders extends HasContext, HasHeaders {
```
maybe we can add an additional method here `copyContextAndHeadersFrom`? and use it everywhere so we don't forget one or the other?

spinscale replied:
fixed... will have to wait for java8 and default methods to have this everywhere...

uboness replied:
yea.. that'd be great to have

uboness commented on src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java:
```diff
@@ -32,7 +34,7 @@
  * Provides a cache key based on its content that can be used to cache shard level response.
  */
-public interface ShardSearchRequest {
+public interface ShardSearchRequest extends HasContext, HasHeaders {
```
`implements HasContextAndHeaders`?

spinscale replied:
fixed

martijnvg commented on src/main/java/org/elasticsearch/percolator/PercolateContext.java:
```diff
@@ -693,4 +690,79 @@ public InnerHitsContext innerHits() {
         throw new UnsupportedOperationException();
     }
 
+    // TODO needs real implementation?
```
I don't think so. The percolator isn't executing requests during query parsing.
Maybe add `assert false : "percolator should get here..."? So if in the future this does happen to be the case the test will fail and then we know about it

spinscale replied:
did this when putting something in context or the headers or when calling the `copy*()` methods... doesnt really matter for reading methods IMO

spinscale commented:
@martijnvg @uboness thx for the comments, incorporated

uboness commented:
LGTM on my side... @jaymode wanna have a final review?

martijnvg commented:
LGTM from my side too.

jaymode commented:
LGTM as well

javanna commented on src/test/java/org/elasticsearch/transport/ContextAndHeaderTransportTests.java:
```diff
@@ -0,0 +1,438 @@
+    }
+
+    @Test
+    public void testThatPercolatingExistingDocumentGetRequestContainsContextAndHeaders() throws Exception {
```
nice method name :D
